dialog/file.go

```diff
@@ -0,0 +1,244 @@
+		widget.NewButton("Cancel", func() {
+			f.win.Hide()
+			if f.callback != nil {
+				f.callback("")
+			}
+		}),
+		f.open)
+	footer := fyne.NewContainerWithLayout(layout.NewBorderLayout(nil, nil, nil, buttons),
+		buttons, f.fileName)
+
+	f.files = fyne.NewContainerWithLayout(layout.NewFixedGridLayout(fyne.NewSize(fileIconSize+16,
+		fileIconSize+theme.Padding()+fileTextSize)),
+	)
+
+	f.breadcrumb = widget.NewHBox()
+	scrollBread := widget.NewScrollContainer(f.breadcrumb)
+	body := fyne.NewContainerWithLayout(layout.NewBorderLayout(scrollBread, nil, nil, nil),
+		scrollBread, widget.NewScrollContainer(f.files))
+	header := widget.NewLabelWithStyle(label+" File", fyne.TextAlignLeading, fyne.TextStyle{Bold: true})
+	favourites := widget.NewGroup("Favourites", f.loadFavourites()...)
+	return fyne.NewContainerWithLayout(layout.NewBorderLayout(header, footer, favourites, nil),
+		favourites, header, footer, body)
+}
+
+func (f *fileDialog) loadFavourites() []fyne.CanvasObject {
+	home, _ := os.UserHomeDir()
+	places := []fyne.CanvasObject{
+		widget.NewButton("Home", func() {
+			f.setDirectory(home)
+		}),
+		widget.NewButton("Documents", func() {
+			f.setDirectory(filepath.Join(home, "Documents"))
+		}),
+		widget.NewButton("Downloads", func() {
+			f.setDirectory(filepath.Join(home, "Downloads"))
+		}),
+	}
+
+	places = append(places, f.loadPlaces()...)
+	return places
+}
+
+func (f *fileDialog) setFileDir(dir *fileIcon) {
+	f.setDirectory(dir.path)
+}
+
+func (f *fileDialog) setDirectory(dir string) {
+	f.setFile(nil)
+	f.dir = dir
+
+	f.breadcrumb.Children = nil
+	buildDir := filepath.VolumeName(dir)
+	for i, d := range strings.Split(dir, string(filepath.Separator)) {
+		if d == "" {
+			if i > 0 { // what we get if we split "/"
+				break
+			}
+			buildDir = "/"
+			d = "/"
+		} else if i > 0 {
+			buildDir = filepath.Join(buildDir, d)
+		} else {
+			d = buildDir
+			buildDir = d + string(os.PathSeparator)
+		}
+
+		newDir := buildDir
+		f.breadcrumb.Append(
+			widget.NewButton(d, func() {
+				f.setDirectory(newDir)
+			}),
+		)
+	}
+
+	f.refreshDir(dir)
+}
+
+func (f *fileDialog) refreshDir(dir string) {
+	f.files.Objects = nil
+
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		fyne.LogError("Unable to read path "+dir, err)
+		return
+	}
+
+	var icons []fyne.CanvasObject
+	parent := filepath.Dir(dir)
+	if parent != dir {
+		icons = append(icons, f.newFileIcon(theme.FolderOpenIcon(), filepath.Dir(dir), f.setFileDir))
+	}
+	for _, file := range files {
+		if isHidden(file.Name(), dir) {
+			continue
+		}
+
+		itemPath := filepath.Join(dir, file.Name())
+		if file.IsDir() {
+			icons = append(icons, f.newFileIcon(theme.FolderIcon(), itemPath, f.setFileDir))
+		} else {
+			icons = append(icons, f.newFileIcon(theme.FileIcon(), itemPath, f.setFile))
+		}
+	}
+
+	f.files.Objects = icons
+	f.files.Refresh()
+}
+
+func (f *fileDialog) setFile(file *fileIcon) {
+	if f.current != nil {
+		f.current.current = false
```

Comment:
`.current.current` stutters. Even with the later proposed change of the latter to `isCurrent` the resulting `.current.isCurrent` wouldn't read good. A current thing is supposed to be current.

I suggest to rename `f.current` to something more expressive. Maybe `f.selection`?

The name of `setFile` should be adjusted accordingly (e.g. `setSelection`).

And the name of the type `fileIcon` is misleading. How about `fileDialogItem`?